dbms/src/DataStreams/ProtobufRowInputStream.cpp

```diff
@@ -0,0 +1,175 @@
+#include <Common/config.h>
+/* #if USE_PROTOBUF */
+
+#include <fcntl.h>
+// TODO: remove
+#include <iostream>
+
+#include <Core/ColumnWithTypeAndName.h>
+#include <Common/escapeForFileName.h>
+#include <IO/copyData.h>
+#include <IO/ReadBuffer.h>
+#include <IO/WriteBufferFromString.h>
+
+/* #include <google/protobuf/compiler/parser.h> */
+/* #include <google/protobuf/descriptor.h> */
+/* #include <google/protobuf/dynamic_message.h> */
+#include <google/protobuf/io/zero_copy_stream.h>
+#include <google/protobuf/io/zero_copy_stream_impl.h>
+/* #include <google/protobuf/io/tokenizer.h> */
+
+#include <DataStreams/ProtobufRowInputStream.h>
+
+namespace DB
+{
+
+using namespace google::protobuf;
+using namespace google::protobuf::io;
+using namespace google::protobuf::compiler;
+
+namespace {
+
+static String getSchemaPath(const String & schema_dir, const String & schema_file)
+{
+    return schema_dir + escapeForFileName(schema_file) + ".proto";
+}
+
+static std::unordered_map<std::string, const FieldDescriptor *> name_to_field;
+
+static const std::unordered_map<FieldDescriptor::CppType, const std::string> protobuf_type_to_column_type = {
+    // TODO: what if numeric values are always smaller than int32/64?
+    {FieldDescriptor::CPPTYPE_INT32,  "Int32"},
+    {FieldDescriptor::CPPTYPE_INT64,  "Int64"},
+    {FieldDescriptor::CPPTYPE_UINT32, "UInt32"},
+    {FieldDescriptor::CPPTYPE_UINT64, "UInt64"},
+    {FieldDescriptor::CPPTYPE_DOUBLE, "Double"},
+    {FieldDescriptor::CPPTYPE_FLOAT,  "Float"},
+    {FieldDescriptor::CPPTYPE_BOOL,   "UInt8"},
+    {FieldDescriptor::CPPTYPE_STRING, "String"}
+    // TODO:
+    /* {Enum, FieldDescriptor::CPPTYPE_ENUM}, */
+    /* FieldDescriptor::CPPTYPE_MESSAGE, */
+    /* FieldDescriptor::MAX_CPPTYPE */
+};
+
+}
+
+ProtobufRowInputStream::ProtobufRowInputStream(
+    ReadBuffer & istr_, const Block & header_,
+    const String & schema_dir, const String & schema_file, const String & root_obj
+) : istr(istr_), header(header_)
+{
+    String schema_path = getSchemaPath(schema_dir, schema_file);
+    // TODO: remove
+    std::cerr << "schema_path: " << schema_path << std::endl;
+    std::cerr << "root obj: " << root_obj << std::endl;
+
+    int fd = open(schema_path.c_str(), O_RDONLY);
```

Comment:
It is better to use high-level interface.
If an exception is thrown below, the open file will remain.